---
# title: "Visuzlize"
# author: "Kasthuri Kannan"
# date: "Sept 07, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---

### Data science fundamentals 01: Visualize and explore RNA-seq data

---

Data exploration is the art of looking at the data, rapidly generating hypotheses, quickly testing them, then repeating again and again, and again. The goal of data exploration is to generate many promising leads that we can later explore in more depth. Visualization is key for data exploration, but not sufficient. We need data transformation and exploratory data analysis which we will cover in this section. Modeling is important too but we will explore it later.

<br>
<span style="color:blue">**Classwork/Homework**</span>: Get familiar with the RNA-seq data - we will be extensively using them.

---

#### Prerequisites

Install package tidyverse and load it.

```{r eval=FALSE}
  # install.packages("tidyverse")
  library(tidyverse)
```

---

Recall data frames. Data frames are lists of lists. We will first import the RNA-seq data and make it into data frames for visualization.

---

#### Creating a `ggplot`

We will first import the RNA-seq data set in RStudio:

```{r eval=TRUE, message=FALSE}
  # Load the package RCurl
  library(RCurl)
  # Import the RNAseq data set from GitHub; break the string into two for readability
  # (Please note this readability aspect very carefully)
  URL_text_1 <- "https://raw.githubusercontent.com/kannan-kasthuri/kannan-kasthuri.github.io"
  URL_text_2 <- "/master/Workshops/DS_Transcriptomics/Data/marked_data_7000_genes.txt"
  # Paste it to constitute a single URL and import the data
  URL <- paste(URL_text_1,URL_text_2, sep="")
  imported.RNA.seq <- read.csv(text=getURL(URL), sep="\t", header = TRUE, stringsAsFactors = FALSE, row.names = 1)
  # Get the gene names
  gene.names <- row.names(imported.RNA.seq)
  # Append "Patiend ID (PID)" variable to make "our" data frame
  column.names <- append("PID",gene.names)
  column.names
  # Get only the gene expression matrix
  expression <- as.numeric(t(imported.RNA.seq[c(2:6917),c(1:52)]))
  # Get the group values
  group <- t(imported.RNA.seq[1,])
```

```{r eval=TRUE, message=FALSE}
 a <- t(imported.RNA.seq )
 a$PID <- row.names(a)
 
# Make our data frame
#df <- data.frame(matrix(ncol = 6919, nrow = 52))
# The first column are the patient ids
#df[,] <- column.names
# df
# The next two columns are Patient ID (PID) and Group. The rest are official gene names
# column.names <- append(c("PID","Group"),merged.data$SYMBOL)
# colnames(df) <- column.names
# Assign the group information
# df[,2] <- group[,1]
# The rest of the matrices are expression values
# expression <- as.numeric(t(merged.data[c(1:6678),c(2:53)]))
# df[c(1:52),c(3:6680)] <- expression
```


We can now use these official symbol for the rest of the analysis.

We create a `ggplot` as follows:

```{r eval=TRUE, message=FALSE, warning=FALSE}
  # Load the tidyverse library
  library(tidyverse)
  # Make a ggplot
  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(DCN), y = log(EPCAM)))
```

With ggplot2, we plot with the function `ggplot()`. `ggplot()` creates a coordinate system that we can add layers to. The first argument of `ggplot()` is the dataset to use in the graph. So `ggplot(data = df)` creates an empty graph.

We add one or more layers to `ggplot()`. The function `geom_point()` adds a layer of points to the plot, which creates a scatterplot. ggplot2 comes with many geom functions that each add a different type of layer to a plot.  
Each geom function in ggplot2 takes a mapping argument. This defines how variables in our dataset are mapped to visual properties. The mapping argument is always paired with `aes()`, and the x and y arguments of `aes()` specify which variables to map to the x and y axes. ggplot2 looks for the mapped variable in the data argument, in this case, our dataframe `df`.

The general graphing template is given by

```{r eval=FALSE}
  ggplot(data = <DATA>) + 
  <GEOM_FUNCTION>(mapping = aes(<MAPPINGS>))
```

<br>
<span style="color:blue">**Classwork/Homework**</span>: 

1. Run `ggplot(data = df)`. What do you see?
2. How many rows are in df? How many columns?
3. What are the expression values for IDH1 in the dataset? 
4. Make a scatterplot of IDH1 vs PTEN.

---

#### Aesthetic mappings

We can add a third variable, like Group, to a two dimensional scatterplot by mapping it to an **aesthetic**. An aesthetic is a visual property of the objects in the plot. Aesthetics include things like the size, the shape, or the color of the points. We can display a point in different ways by changing the values of its aesthetic properties. 

We can convey information about the data by mapping the aesthetics in the plot to the variables in df dataset. For example, we can map the colors of our points to reveal the group status of each person.

```{r eval=TRUE, message=FALSE, warning=FALSE}
  # Make a ggplot with asthetic color for the variable Group
  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(DCN), y = log(EPCAM), color=Group))
```

To map an aesthetic to a variable, associate the name of the aesthetic to the name of the variable inside `aes()`. ggplot2 will automatically assign a unique level of the aesthetic (here a unique color) to each unique value of the variable, a process known as _scaling_. ggplot2 will also add a legend that explains which levels correspond to which values.

The colors reveal many things about this data set. We see that the _Luminal_ group has relatively high expression for _FGFR4_ compared to _Normal-like_ group.

This data set does not have other categorical variables. To demonstrate the power of `ggplot2` we create another table with categories 'Age-group' and 'Chemo' status, that shows the age group of the patient and whether they underwent chemotherapy. We import the data set just like how we did previously.

```{r eval=TRUE, message=FALSE}
  # Load the package RCurl
  library(RCurl)
  # Import the RNAseq data set from GitHub; break the string into two for readability
  # (Please note this readability aspect very carefully)
  URL_text_1 <- "https://raw.githubusercontent.com/kannan-kasthuri/kannan-kasthuri.github.io"
  URL_text_2 <- "/master/Workshops/DS_Transcriptomics/Data/marked_data_7000_genes_other_attributes.txt"
  # Paste it to constitute a single URL and import the data
  URL <- paste(URL_text_1,URL_text_2, sep="")
  imported.RNA.seq_other_factors <- read.csv(text=getURL(URL), sep="\t", header = TRUE, stringsAsFactors = FALSE, row.names = 1)
  # Get the Age Group and Chemo status info
  Age_group <- imported.RNA.seq_other_factors[2,]
  Chemo_status <- imported.RNA.seq_other_factors[3,]
  df$AGEGROUP <- t(Age_group)
  df$CHEMO <- t(Chemo_status)
```

In the above example, we mapped `Group` to the color aesthetic, but we could have mapped AGEGROUP/CHEMO variables to the size aesthetic in the same way. In this case, the exact size of each point would reveal its AGEGROUP affiliation. We get a warning here, because mapping an unordered variable (class) to an ordered aesthetic (size) is not a good idea.

```{r eval=TRUE, message=FALSE, warning=FALSE}
  # Make a ggplot with asthetic size for the variable AGEGROUP
  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(TP53), y = log(FGFR4), size=AGEGROUP))
```

Or we could have mapped AGEGROUP to the alpha aesthetic, which controls the transparency of the points: 

```{r eval=TRUE, message=FALSE, warning=FALSE}
  # Make a ggplot with asthetic alpha for the variable AGEGROUP
  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(FGFR4), y = log(IDH1), alpha=AGEGROUP))
```


or the shape of the points:

```{r eval=TRUE, message=FALSE, warning=FALSE}
  # Make a ggplot with asthetic shape for the variable AGEGROUP
  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(TP53), y = log(PDE4DIP), shape=AGEGROUP))
```

**Note**: ggplot2 will only use six shapes at a time. By default, additional groups will go unplotted when we use the shape aesthetic.

For each aesthetic, we use `aes()` to associate the name of the aesthetic with a variable to display. The `aes()` function gathers together each of the aesthetic mappings used by a layer and passes them to the layer’s mapping argument. The syntax highlights a useful insight about x and y: the x and y locations of a point are themselves aesthetics, visual properties that we can map to variables to display information about the data.

Once we map an aesthetic, ggplot2 takes care of the rest. It selects a reasonable scale to use with the aesthetic, and it constructs a legend that explains the mapping between levels and values. For x and y aesthetics, ggplot2 does not create a legend, but it creates an axis line with tick marks and a label. The axis line acts as a legend; it explains the mapping between locations and values.

We can also set the aesthetic properties of our geom manually. For example, we can make all of the points in our plot red:

```{r eval=TRUE, message=FALSE, warning=FALSE}
  # Make a ggplot with color red
  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(DROSHA), y = log(SARS)), color="red")
```


<br>
<span style="color:blue">**Classwork/Homework**</span>: 

1. What’s gone wrong with this code? Why are the points not blue?

```{r eval=FALSE, message=FALSE, warning=FALSE}
  # Make a ggplot with color blue
  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(DROSHA), y = log(SARS), color="blue"))
```

2. Which variables in df are categorical? Which variables are continuous? How can we see this information?
3. Map a continuous variable to color, size, and shape. How do these aesthetics behave differently for categorical vs. continuous variables?
4. What happens if we map the same variable to multiple aesthetics?
5. What does the stroke aesthetic do? What shapes does it work with? 
6.What happens if we map an aesthetic to something other than a variable name, like aes(colour = A1C < 5)?

---

#### Facets

One way to add additional variables is with aesthetics. Another way, particularly useful for categorical variables, is to split plots into what are known as **facets**, subplots that each display one subset of the data.

To facet a plot by a single variable, the function `facet_wrap()` is useful. The first argument of `facet_wrap()` should be a formula, which we create with ~ followed by a variable name (here “formula” is the name of a data structure in R, not a synonym for “equation”). The variable that we pass to `facet_wrap()` should be discrete.


```{r eval=TRUE, warning=FALSE}
  # Make a ggplot with facets
  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(FGFR4), y = log(IDH1))) + 
  facet_wrap(~ Group, nrow = 2)
```

To facet our plot on the combination of two variables, we should add `facet_grid()` to our plot call. The first argument of `facet_grid()` is also a formula. This time the formula should contain two variable names separated by a ~.

```{r eval=TRUE, warning=FALSE}
  # Make a ggplot with facet grid - Group vs AGEGROUP
  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(DROSHA), y = log(PDE4DIP))) + 
  facet_grid(Group ~ as.vector(AGEGROUP))
```

We see that there are no _Normal-like_ in the age group 51-60, 31-40 etc. However, plotting the facet grid with AGEGROUP vs CHEMO, reveals,

```{r eval=TRUE, warning=FALSE}
  # Make a ggplot with facet grid - AGEGROUP vs CHEMO
  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(TP53), y = log(FGFR4))) + 
  facet_grid(as.vector(AGEGROUP) ~ as.vector(CHEMO))
```

there are hardly people in the age group 51-60 who had chemo, while there are lots of them in 41-50 group who did not receive chemo.

However, inspecting the age group 41-50, we see the P53 expressions are larger in this group who did not have chemo. This is an example of **directed insight** obtained through visualization. 

<br>
<span style="color:blue">**Classwork/Homework**</span>: 

1. What happens if you facet on a continuous variable?
2. What plots does the following code make? What does `.` do?

```{r eval=FALSE}
  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(TP53), y = log(FGFR4))) +
  facet_grid(as.vector(AGEGROUP) ~ .)

  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(PTEN), y = log(IDH1))) +
  facet_grid(. ~ as.vector(chemo))
```

3. Derive directed insights from the above plots.
4. Take the first faceted plot in this section:
```{r eval=FALSE, warning=FALSE}
  # Make a ggplot with facets
  ggplot(data = df) + 
  geom_point(mapping = aes(x = log(FGFR4), y = log(IDH1))) + 
  facet_wrap(~ Group, nrow = 2)
```

What are the advantages to using faceting instead of the colour aesthetic? What are the disadvantages? How might the balance change if you had a larger dataset?

5. Read `?facet_wrap`. What does nrow do? What does ncol do? What other options control the layout of the individual panels? Why doesn’t `facet_grid()` have `nrow` and `ncol` argument?
6. When using `facet_grid()` you should usually put the variable with more unique levels in the columns. Why?

---

#### Geometric objects

How are these two plots similar?

```{r eval=TRUE, warning=FALSE, message=FALSE, echo=FALSE}
  # Plot FGFR4 vs TP53, coloring Group
  ggplot(data = df) + 
    geom_point(mapping = aes(x = FGFR4, y = TP53, color=Group)) + 
    xlim(c(0,13)) + ylim(c(0,10))
```

```{r eval=TRUE, warning=FALSE, message=FALSE, echo=FALSE}
  # Smooth FGFR4 vs TP53, coloring Group
  ggplot(data = df) + 
    geom_smooth(mapping = aes(x = FGFR4, y = TP53, color=Group)) + 
    xlim(c(0,13)) + ylim(c(0,10))
```

Both plots contain the same x variable, the same y variable, and both describe the same data. But the plots are not identical. Each plot uses a different visual object to represent the data. In ggplot2 syntax, they're called **geoms**.

A **geom** is the geometrical object that a plot uses to represent data. 

Plots are often described by the type of geom that the plot uses. For example, bar charts use bar geoms, line charts use line geoms, boxplots use boxplot geoms, and so on. Scatterplots break the trend; they use the point geom. We can use different geoms to plot the same data. The plot on top uses the point geom, and the plot in the bottom uses the smooth geom, a smooth line fitted to the data.

Here is the point geom code:

```{r eval=FALSE, warning=FALSE, message=FALSE, echo=TRUE, results="hide"}
  # Plot point geom (scatter plot) FGFR4 vs TP53, coloring Group
  ggplot(data = df) + 
  geom_point(mapping = aes(x = FGFR4, y = TP53, color=Group)) + 
  xlim(c(0,13)) + ylim(c(0,10))
```

And here is the line geom code:

```{r eval=FALSE, warning=FALSE, message=FALSE, echo=TRUE, results="hide"}
  # Plot smooth geom FGFR4 vs TP53, coloring Group
  ggplot(data = df) + 
  geom_smooth(mapping = aes(x = FGFR4, y = TP53, color=Group)) + 
  xlim(c(0,13)) + ylim(c(0,10))
```

<br>
<span style="color:blue">**Classwork/Homework**</span>: Try `geom_smmoth` on different variables in df data frame.

Every geom function in ggplot2 takes a mapping argument. 

However, not every aesthetic works with every geom. We could set the shape of a point, but we cannot set the “shape” of a line. 

On the other hand, we could set the color/linetype of a line. 

`geom_smooth()` will draw a different line, with a different color for male and female, for each unique value of the variable that we map to color. One can also specify `linetype` instead of color.

In fact, we can over lay two geoms on the **same** plot - super cool, isn't it (compare it with `abline()` function):

```{r eval=TRUE, warning=FALSE, message=FALSE, echo=TRUE}
  # Plot smooth and point geom in the same plot 
  ggplot(data = df) + 
  geom_smooth(mapping = aes(x = FGFR4, y = TP53, color=Group)) + 
  xlim(c(0,13)) + ylim(c(0,10)) +
  geom_point(mapping = aes(x = FGFR4, y = TP53, color=Group))
```

**Note**: This introduces some duplication in our code. Imagine if we wanted to change the y-axis to display IDH1 values instead of TP53. We need to change the variable in two places, and we might forget to update one. We can avoid this type of repetition by passing a set of mappings to the `ggplot()` function. ggplot2 will treat these mappings as global mappings that apply to each geom in the graph. In other words, this code will produce the same plot as the previous code:

```{r eval=TRUE, warning=FALSE, message=FALSE, echo=TRUE}
  # Plot smooth and point geom in the same plot 
  ggplot(data = df, mapping = aes(x = FGFR4, y = TP53, color=Group)) + 
  xlim(c(0,13)) + ylim(c(0,10)) +
  geom_smooth() + 
  geom_point()
```

If we place mappings in a geom function, ggplot2 will treat them as local mappings for the layer. 

It will use these mappings to extend or overwrite the global mappings for that layer only. 

This makes it possible to display different aesthetics in different layers.

```{r eval=TRUE, warning=FALSE, message=FALSE, echo=TRUE}
  # Plot smooth and point geom in the same plot 
  ggplot(data = df, mapping = aes(x = FGFR4, y = TP53)) + 
  xlim(c(0,13)) + ylim(c(0,10)) +
  geom_smooth(mapping = aes(color = Group)) + 
  geom_point() 
```

**Note**: One can use this idea for plotting _different_ data to different layers. For example, the geom `geom_point` can be specified to select AGEGROUP variable consisting only people given chemo, for instance.

ggplot2 provides over 30 geoms, and extension packages provide even more (see https://www.ggplot2-exts.org for a sampling). The best way to get a comprehensive overview is the ggplot2 cheatsheet, which you can find at http://rstudio.com/cheatsheets. To learn more about any single geom, use help: ?geom_smooth.

<br>
<span style="color:blue">**Classwork/Homework**</span>: 

1. What does the `group` argument in aesthetics in the line geom do? Try it in an example and report the result.
2. What geom would you use to draw a line chart? A boxplot? A histogram? An area chart?
3. What does `show.legend = FALSE` do? What happens if you remove it?
4. What does the `se` argument to `geom_smooth()` do?
5. Read sections 3.7, 3.8 in the book [R for Data Science](http://r4ds.had.co.nz/). For each code example (ignore code in exercises) given in these sections, you need to provide similar examples using `df` data sets. Also, change the documentation to reflect your demo.

<br>

#### Selected materials and references

[R for Data Science - Explore Part](https://cran.r-project.org/doc/manuals/R-intro.pdf)

---